ARROW-15573: [Java][Doc] Document Apache Arrow memory management (#12603)
apache/arrow: docs/source/java/memory.rst

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=================
Memory Management
=================

.. contents::

The memory modules contain all the functionality that Arrow uses to manage memory (allocation and deallocation).
This section will introduce you to the major concepts in Java’s memory management:

* `ArrowBuf`_

> * `Reference counting`_

Reviewer: Let's not link to Netty's docs right off the bat, I think that's a little confusing without context

Author: Deleted

* `BufferAllocator`_

Getting Started
===============

Arrow's memory management is built around the needs of the columnar format and using off-heap memory.
Also, it is its own independent implementation, and does not wrap the C++ implementation.

> Arrow offer a high level of abstraction providing several access APIs to read/write data into a direct memory.

Reviewer: ```suggestion
Arrow offers a high level of abstraction providing several access APIs to read/write data into a direct memory.
```

> Arrow provides multiple modules, but users only need two of them:

Reviewer: "Arrow provides multiple modules: the core interfaces, and implementations of the interfaces. Users need the core interfaces, and exactly one of the implementations."

Author: Updated

* ``Memory Core``: Provides the interfaces used by the Arrow libraries and applications.
* ``Memory Netty``: An implementation of the memory interfaces based on the `Netty`_ library.
* ``Memory Unsafe``: An implementation of the memory interfaces based on the `sun.misc.Unsafe`_ library.

ArrowBuf
========

> ArrowBuf represents a single, contiguous allocation of direct memory. It consists of an address and a length,

Reviewer: nit: link to the API docs again here

Reviewer: We need more context here in general. We should link "direct memory" to https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/ByteBuffer.html since it explains direct byte buffers.

We should also briefly summarize what direct memory is, and why Arrow uses it.

Author: Added

and provides low-level interfaces for working with the contents, similar to ByteBuffer.

Unlike (Direct)ByteBuffer, it has reference counting built in (see the next section).

Reference counting
==================

> 

Reviewer: More context here as well. "Direct memory is more expensive to allocate and deallocate, so allocators pool/cache allocated buffers." (We could move Allocator before this section.) "Arrow uses manual reference counting to track whether a buffer is in use, or can be deallocated or returned to the allocator's pool. This simply means that each buffer has a counter keeping track of the number of references to the buffer, and code is responsible for properly incrementing/decrementing the counter as the buffer is used. (Some programming languages offer this as a general memory management strategy, albeit usually hidden from the user to some extent.) 

In Arrow, each buffer has an associated ReferenceManager that tracks the reference count, which can be retrieved with ArrowBuf.getReferenceManager(). The reference count can be updated with ReferenceManager.release and ReferenceManager.retain. Of course, this is tedious and error-prone, so usually, instead of directly working with buffers, we use higher-level APIs like ValueVector. Such classes generally implement Closeable/AutoCloseable and will automatically decrement the reference count when closed."

also feel free to rewrite things, I'm just trying to sketch out the amount of context I think we should provide

Author: Thank you

Is a technique to help computer programs manage memory. Tracks the reference/pointers to an object, it increase
+1 or decrease -1 the reference counting between the objects.

If an object ValidityBuffer has a reference with object IntVector, then, IntVector should increase the
reference counting to 1 (0 + 1 = 1), then if at the same time, ValueBuffer has a reference with IntVector,

> then, IntVector should increase the reference counting to 2 (1 + 1 = 2).

Reviewer: Is this the case? As described this sounds like a reference cycle.

Author: Deleted and change by suggested

.. code-block::

    |__ A = Allocator
    |____ B = IntVector (reference count = 2 )
    |____________ ValidityBuffer
    |____________ ValueBuffer
    |____ C = VarcharVector (reference count = 2 )
    |____________ ValidityBuffer
    |____________ ValueBuffer

Base on best practices at some point you are going to close your allocator objects using ``close()`` method,

> allocators check for reference counting and throw an exception if they are in use.

Reviewer: "Allocators implement AutoCloseable as well. In this case, closing the allocator will check that all buffers obtained from the allocator are closed. If not, close() will raise an exception; this helps track memory leaks from unclosed buffers. When combined with child allocators, this is especially useful for ensuring at an independent section of code has fully cleaned up all allocated buffers, while still maintaining a global memory limit through the RootAllocator."

Author: Added

Reference Manager manages the reference counting for the underlying memory chunk.

Allocators
==========

One of the interfaces defined by memory-core is BufferAllocator. This interface collect all the definitions for deal

> with byte buffer allocation.

Reviewer: "The BufferAllocator interface deals with allocating ArrowBufs for the application."

Author: Thanks, changed

> The concrete implementation of the allocator is Root Allocator. Applications should generally
> create one allocator at the start of the program..

Reviewer: ```suggestion
The concrete implementation of the allocator is RootAllocator. Applications should generally
create one allocator at the start of the program, and use it through the BufferAllocator interface.
```

Arrow provides a tree-based model for memory allocation. The RootAllocator is created first,

> then all allocators are created as children of that allocator. The RootAllocator is responsible

Reviewer: Add the method used to create child allocators. 

Mention that allocators have a memory limit. The RootAllocator then effectively sets the program-wide memory limit. One of the uses of child allocators is to set a lower temporary limit for one section of the code. Also, child allocators can be named; this makes it easier to tell where an ArrowBuf came from during debugging.

Author: Added

for being the master bookkeeper for memory allocations.

> As an example of child allocator consider `Flight Client`_ creation.

Reviewer: IMO, pointing into internal implementation details does not make for a good example. We should add and link to a cookbook example.

Author: Deleted

Memory Modules
==============

Applications should depend on memory-core and one of the two implementations,

> else an exception will be raised at runtime.

Reviewer: This is redundant with above.

Author: Deleted

Development Guidelines
======================

* Use the BufferAllocator interface in APIs instead of RootAllocator.

> * Applications should generally create one allocator at the start of the program.

Reviewer: ```suggestion
* Applications should generally create one RootAllocator at the start of the program.
```

* Remember to close() allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.

> * Allocators will check for outstanding memory allocations when closed, and throw an exception if there are allocated buffers, this helps detect memory leaks.

Reviewer: This isn't a guideline IMO (it should be something specific and concrete that developers should do, not a description of the Arrow implementation)

Author: Deleted

> * Allocators have a debug mode, that makes it easier to figure out where a leak originated (Consider to add this parameter to your application: -Darrow.memory.debug.allocator=true)

Reviewer: What does the debug mode do? Also, I think this can be a subsection of the Allocators section, since we should expand on it

Author: Added Debugging Memory Leaks/Allocation

> * Arrow modules use logback to collect logs configure that properly to see your logs (create logback-test.xml file on resources folder).

Reviewer: I think this can also go as part of a section about debugging memory leaks/allocations

Author: Added Debugging Memory Leaks/Allocation

.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=================
Memory Management
=================

.. contents::

The memory modules contain all the functionality that Arrow uses to manage memory (allocation and deallocation).
This section will introduce you to the major concepts in Java’s memory management:

* `BufferAllocator`_
* `ArrowBuf`_
* `Reference counting`_

Getting Started
===============

Arrow's memory management is built around the needs of the columnar format and using off-heap memory.
Also, it is its own independent implementation, and does not wrap the C++ implementation.

> Arrow offers a high level of abstraction providing several access APIs to read/write data into a direct memory.

Author: Deleted

Arrow provides multiple modules: the core interfaces, and implementations of the interfaces.
Users need the core interfaces, and exactly one of the implementations.

> * ``Memory Core``: Provides the interfaces used by the Arrow libraries and applications.

Author: Updated

* ``Memory Netty``: An implementation of the memory interfaces based on the `Netty`_ library.
* ``Memory Unsafe``: An implementation of the memory interfaces based on the `sun.misc.Unsafe`_ library.

BufferAllocator
===============

> The BufferAllocator interface deals with allocating ArrowBufs for the application.

Author: Added

The concrete implementation of the allocator is RootAllocator. Applications should generally create one RootAllocator at the
start of the program, and use it through the BufferAllocator interface. Allocators have a memory limit. The RootAllocator
sets the program-wide memory limit. The RootAllocator is responsible for being the master bookkeeper for memory allocations.

Arrow provides a tree-based model for memory allocation. The RootAllocator is created first, then all allocators
are created as children ``BufferAllocator.newChildAllocator`` of that allocator.

One of the uses of child allocators is to set a lower temporary limit for one section of the code. Also, child

> allocators can be named; this makes it easier to tell where an ArrowBuf came from during debugging.

Reviewer: Hmm, I feel like we can still introduce this better.

"The concrete implementation of the BufferAllocator interface is RootAllocator. (Link to the Javadocs.) Applications should generally create one RootAllocator at the start of the program, and use it through the BufferAllocator interface. Allocators implement AutoCloseable and must be closed after the application is done with them; this will check that all outstanding memory has been freed (see the next section).

Arrow provides a tree-based model for memory allocation. The RootAllocator is created first, then more allocators
are created as children of an existing allocator via newChildAllocator. (Link to Javadocs.) When creating a RootAllocator or a child allocator, a memory limit is provided, and when allocating memory, the limit is checked. Furthermore, when allocating memory from a child allocator, those allocations are also reflected in all parent allocators. Hence, the RootAllocator effectively sets the program-wide memory limit, and serves as the master bookkeeper for all memory allocations.

Child allocators are not strictly required, but can help better organize code. For instance, a lower memory limit can be set for a particular section of code. When the allocator is closed, it then checks that that section didn't leak any memory. And child allocators can be named, which makes it easier to tell where an ArrowBuf came from during debugging."

Author: Thanks, updated

> ArrowBuf

Author: Changed

========

ArrowBuf represents a single, contiguous allocation of `Direct Memory`_. It consists of an address and a length,
and provides low-level interfaces for working with the contents, similar to ByteBuffer.

The objects created using ``Direct Memory`` take advantage of native executions and it is decided natively by the JVM. Arrow

> offer efficient memory operations base on this Direct Memory implementation (`see section below for detailed reasons of use`).

Author: Updated

Unlike (Direct)ByteBuffer, it has reference counting built in (`see the next section`).

Reference counting
==================

Direct memory involve more activities than allocate and deallocate because allocators (thru pool/cache)

> allocate buffers (ArrowBuf).

Reviewer: Direct memory is more expensive to allocate and deallocate. That's why allocators pool or cache direct buffers. The caching is not the _reason_ why it's expensive, it's the _consequence_.

Author: Thank you, added on my notes. Updated

> Arrow uses manual reference counting to track whether a buffer is in use, or can be deallocated or returned

Author: Added

to the allocator's pool. This simply means that each buffer has a counter keeping track of the number of references to

> this buffer, and end user is responsible for properly incrementing/decrementing the counter according the buffer is used.

Reviewer: ```suggestion
the buffer, and the user is responsible for properly incrementing/decrementing the counter as the buffer is used.
```

In Arrow, each ArrowBuf has an associated ReferenceManager that tracks the reference count, which can be retrieved

> with ArrowBuf.getReferenceManager(). The reference count can be updated with ``ReferenceManager.release`` and

Reviewer: Link to Javadocs.

Author: Added

Reviewer: Was this added?

Author: It was added to ReferenceManager, just added also to release and retain

``ReferenceManager.retain``.

Of course, this is tedious and error-prone, so usually, instead of directly working with buffers, we should use
higher-level APIs like ValueVector. Such classes generally implement Closeable/AutoCloseable and will automatically

> decrement the reference count when closed method.

Reviewer: ```suggestion
decrement the reference count when closed.
```

.. code-block::

    |__ A = Allocator

>     |____ B = IntVector (reference count = 2 )

Reviewer: This is wrong, no? Buffers have a reference count, not vectors. And ValidityBuffer and ValueBuffer aren't classes.

    |____________ ValidityBuffer
    |____________ ValueBuffer
    |____ C = VarcharVector (reference count = 2 )
    |____________ ValidityBuffer
    |____________ ValueBuffer

Allocators implement AutoCloseable as well. In this case, closing the allocator will check that all buffers
obtained from the allocator are closed. If not, ``close()`` method will raise an exception; this helps track
memory leaks from unclosed buffers.

> As you see reference counting needs to be handled properly by us, if at some point you need to ensuring that an
> independent section of code has `fully cleaned up all allocated buffers while still maintaining a global memory limit
> through the RootAllocator`, well ``BufferAllocator.newChildAllocator`` is what you should use.

Reviewer: ```suggestion
As you see, reference counting needs to be handled carefully. To ensure that an
independent section of code has fully cleaned up all allocated buffers, use a new child allocator.
```

Reason To Use Direct Memory
===========================

* When `writing an ArrowBuf`_ we use the direct buffer (``nioBuffer()`` returns a DirectByteBuffer) and the JVM `will attempt to avoid copying the buffer's content to (or from) an intermediate buffer`_ so it makes I/O (and hence IPC) faster.

> * We can `directly wrap a native memory address`_ instead of having to copy data for JNI (where in implementing the C Data Interface we can directly create `Java ArrowBufs that directly correspond to the C pointers`_).

Author: Updated

> * Conversely in JNI, we can directly use `Java ArrowBufs in C++`_ without having to copy data.

Author: Updated

So basically #1 is more efficient I/O, and #2/#3 is better integration with JNI code.

Development Guidelines
======================

* Use the BufferAllocator interface in APIs instead of RootAllocator.
* Applications should generally create one RootAllocator at the start of the program.
* Remember to close() allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.

Debugging Memory Leaks/Allocation
=================================

Allocators have a debug mode, that makes it easier to figure out where a leak is originated (Consider to add this
parameter to your application: ``-Darrow.memory.debug.allocator=true``). This parameter enable to create an historical log
about the memory allocation.

Arrow modules use logback to collect logs, configure it properly to see your logs (create ``logback-test.xml`` file on

> resources folder and your project could read that by conventions).

Author: Updated

> This is an example of historical log enabled:

Author: Added snippet code

Logs with debug allocators enabled:

.. code-block:: shell

    11:56:48.944 [main] INFO  o.apache.arrow.memory.BaseAllocator - Debug mode enabled.
    ArrowBuf[2], address:140437894463488, length:4096
    Exception in thread "main" java.lang.IllegalStateException: Allocator[ROOT] closed with outstanding buffers allocated (1).
    Allocator(ROOT) 0/4096/4096/8192 (res/actual/peak/limit)
      child allocators: 0
      ledgers: 1
        ledger[1] allocator: ROOT), isOwning: , size: , references: 1, life: 261438177096661..0, allocatorManager: [, life: ] holds 1 buffers.
            ArrowBuf[2], address:140437894463488, length:4096
      reservations: 0

.. _`BufferAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/BufferAllocator.html
.. _`RootAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/RootAllocator.html
.. _`newChildAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/RootAllocator.html#newChildAllocator-java.lang.String-org.apache.arrow.memory.AllocationListener-long-long-
.. _`ArrowBuf`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/ArrowBuf.html

> .. _`Reference Counting`: https://github.com/apache/arrow/blob/2092e18752a9c0494799493b12eb1830052217a2/java/memory/memory-core/src/main/java/org/apache/arrow/memory/ReferenceManager.java#L30

Reviewer: Wait - I don't think we should link into source files. They don't really provide useful context in this tutorial. (That goes for all the links here.)

Author: Deleted

decrement the reference count when closed.

Allocators implement AutoCloseable as well. In this case, closing the allocator will check that all buffers
obtained from the allocator are closed. If not, ``close()`` method will raise an exception; this helps track
memory leaks from unclosed buffers.

As you see, reference counting needs to be handled carefully. To ensure that an
independent section of code has fully cleaned up all allocated buffers, use a new child allocator.

Development Guidelines
======================

Applications should generally:

* Use the BufferAllocator interface in APIs instead of RootAllocator.
* Create one RootAllocator at the start of the program.
* ``close()`` allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.

Debugging Memory Leaks/Allocation
=================================

Allocators have a debug mode that makes it easier to figure out where a leak is originated.
To enable it, enable assertions with ``-ea`` or set the system property, ``-Darrow.memory.debug.allocator=true``.
When enabled, a log will be kept of allocations.

> Arrow modules define simple logging facade for java SLF4J, configure it properly to see your logs (e.g. Logback/Log4J).

Reviewer: We don't define slf4j.

Reviewer: ```suggestion
Arrow logs some allocation information via SLF4J; configure it properly to see these logs (e.g. via Logback/Apache Log4j).
```

    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.IntVector;

    try (BufferAllocator bufferAllocator = new RootAllocator(Integer.MAX_VALUE)) {
        final int QUANTITY = 5;
        try (IntVector intVector = new IntVector("int-01", bufferAllocator)) {
            intVector.allocateNew(QUANTITY);
            for (int i = 0; i < QUANTITY; i++) {
                intVector.set(i, i);
            }
            intVector.setValueCount(QUANTITY);
        }
        // Fix the next code!, it is only to see the track of allocators when debug is enabled
        IntVector intVectorV = new IntVector("int-02", bufferAllocator);
        intVectorV.allocateNew(QUANTITY);
        for (int i = 0; i < QUANTITY; i++) {
            intVectorV.set(i, i);
        }
        intVectorV.setValueCount(QUANTITY);

        BufferAllocator childAllocator = bufferAllocator.newChildAllocator("child-isolated", 0,
                Integer.MAX_VALUE / 4);
        IntVector intVectorV2 = new IntVector("int-isolated-01", childAllocator);
        intVectorV2.allocateNew(QUANTITY);
        for (int i = 0; i < QUANTITY; i++) {
            intVectorV2.set(i, i);
        }

>     }

Reviewer: We should only use ArrowBuf here, don't use IntVector or anything as they haven't been introduced yet.

Also, it's still not clear to me what the debug mode is doing here. From what I recall, the information it provides is basically only available in a debugger. If there is something the debug mode is doing here, we need to point it out explicitly. Ideally we would compare the information available with and without debug mode.

Author: Added a example with/without use this parameter

Reviewer: Thanks.

I'm pretty sure you can get a stack trace showing where each buffer was allocated, but that may only be in the debugger. I'll try to follow up tomorrow…

Reviewer: I'll submit a jira/PR for this later (there's some issues that need fixing) but I think things are good here as is.

.. code-block::

    15:49:32,755 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback-test.xml] at [file:/Users/java/source/demo/target/classes/logback-test.xml]
    15:49:32,924 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.apache.arrow] to DEBUG
    11:56:48.944 [main] INFO  o.apache.arrow.memory.BaseAllocator - Debug mode enabled.
    Exception in thread "main" java.lang.IllegalStateException: Allocator[ROOT] closed with outstanding child allocators.
    Allocator(ROOT) 0/64/64/2147483647 (res/actual/peak/limit)
      child allocators: 1
        Allocator(child-isolated) 0/32/32/536870911 (res/actual/peak/limit)
          child allocators: 0
          ledgers: 1
            ledger[3] allocator: child-isolated), isOwning: , size: , references: 2, life: 246918908438818..0, allocatorManager: [, life: ] holds 3 buffers.
                ArrowBuf[10], address:140408097079352, length:8
                ArrowBuf[8], address:140408097079328, length:32
                ArrowBuf[9], address:140408097079328, length:24
          reservations: 0
      ledgers: 1

>         ledger[2] allocator: ROOT), isOwning: , size: , references: 2, life: 246tors can be named; this makes it easier to tell where an Arro918906331643..0, allocatorManager: [, life: ] holds 3 buffers.

Reviewer: Seems something got copy-pasted into the middle here?

Author: Deleted

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=================
Memory Management
=================

> The memory management package contains all the memory allocation related items that Arrow uses to manage memory.

Reviewer: This sentence is a little redundant

Reviewer: ```suggestion
The memory modules contain all the functionality that Arrow uses to manage memory (allocation and deallocation).
```

This section will introduce you to the major concepts in Java’s memory management:

> * Allocator

Reviewer: Maybe `BufferAllocator` and link to the Javadoc

Author: Updated

> * Arrowbuf

Reviewer: Maybe `ArrowBuf` and link to the Javadoc

Author: Updated

> 

Reviewer: I would argue that "Reference counting" is a third major concept

Author: Added

.. contents::

Getting Started
===============

Java memory implementation is independently from C++ (not a wrapper around).

> Java memory was implemented considering these specifications: Arrow Columnar Format and Java Off Heap references.

Reviewer: Hmm, the focus here is a little off IMO. Something like this: Arrow's memory management is built around the needs of the columnar format and using off-heap memory. Also, it is its own independent implementation, and does not wrap the C++ implementation.

Reviewer: We should elaborate on how and why Arrow uses direct memory

Author: Added

.. note::

>     Java Memory Data = Data (Columnar mode reference) + Metadata (Flatbuffers serialization reference).

Reviewer: I'm not really sure what this tells us, what is the user learning here?

Author: Deleted

> These are the java memory modules:

Reviewer: Arrow provides multiple modules, but users only need one or two of them:

Author: Updated

> * Memory Core: Core off-heap memory management libraries for Arrow ValueVectors.

Reviewer: ```
``memory-core``: provides the interfaces used by the Arrow libraries and applications.
```

Author: Updated

> * Memory Netty: Netty allocator and utils for allocating memory in Arrow.

Reviewer: ```
``memory-netty``: an implementation of the memory interfaces based on the [Netty](INSERT LINK HERE) library.
```

Author: Updated

> * Memory Unsafe: Allocator and utils for allocating memory in Arrow based on sun.misc.Unsafe.

Reviewer: and ditto

Author: Updated

Allocators
==========

> Memory core module define the next allocators:

Reviewer: Again, for a beginner, I think we can have more explanation, and focus less on just replicating the README 1-1

Reviewer: "One of the interfaces defined by memory-core is BufferAllocator. This class is responsible for..."

Author: Updated

* Buffer Allocator: The public interface application users should be leveraging.

> * Root Allocator: A root allocator for using direct memory. Typically only one created for a JVM.

Reviewer: "The concrete implementation of the allocator. Applications should generally create one allocator at the start of the program." 

The direct memory stuff should be described above, not here.

Also, if we are referencing a class name, then just use the class name (ideally with a link to the docs), e.g. "RootAllocator", else it gets confusing

Author: Updated

Arrow provides a tree-based model for memory allocation. The RootAllocator is created first,
then all allocators are created as children of that allocator. The RootAllocator is responsible

> for being the master bookkeeper for memory allocations.

Reviewer: Why do we want to use child allocators?

Author: Added an example,  but please @lidavidm  if you could help me with more examples or reason to use child allocators.

> Please consider this note on your development:

Reviewer: "Consider these guidelines:"

Author: Updated

> * Use BufferAllocator instead of RootAllocator in your allocator creation.

Reviewer: "Use the BufferAllocator interface in APIs instead of RootAllocator."

Author: Updated

> * Create your allocator inside of a try-with-resources statement.

Reviewer: "Remember to close() allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement. Allocators will check for outstanding memory allocations when closed, and throw an exception if there are allocated buffers; this helps detect memory leaks."

Reviewer: Allocators have a debug mode, IIRC, that makes it easier to figure out where a leak originated

Author: Added

.. code-block:: Java

>     try (BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE) ) { ; }

Reviewer: Java users know how to use try-with-resources. We should assume they are a beginner to Arrow, but are reasonably familiar with Java itself.

Author: Deleted

ArrowBuf
========

> The facade for interacting directly with a chunk of memory.

Reviewer: "ArrowBuf represents a single, contiguous allocation of direct memory. It consists of an address and a length, and provides low-level interfaces for working with the contents, similar to ByteBuffer. Unlike (Direct)ByteBuffer, it has reference counting built in (see the next section)."

Author: Added

Two important instance variables of an ArrowBuf:

* Address: Starting virtual address in the underlying memory chunk that this ArrowBuf has access to.

> * Length: Length (in bytes) in the underlying memory chunk that this ArrowBuf has access to.

Reviewer: I don't think we need to elaborate on this so much.

Author: Deleted

> 

Reviewer: As suggested above we should have a section on reference counting. We should talk about remembering to `close()` objects here

Author: Added

Memory Modules
==============

Memory core define the bases to work with direct memory and the application decided to allocate arrow buffer bases on
the dependency added on your pom.xml (memory-unsafe or memory-netty). If any of these is not added the application raise

> an exception.

Reviewer: This should just be one sentence as part of the list of modules above. "Applications should depend on memory-core and one of the two implementations, else an exception will be raised at runtime."

We generally shouldn't talk about pom.xml. Not all people use Maven.

Author: Updated
